### PR TITLE
Lib tests

### PR DESCRIPTION
**Pull request by AmalieDue**
### Description

Library end-to-end tests have been added which means tests of the complete user-facing functionality.

### Relevant Issues/PRs
Closes #xxx
PR #xxx

### Type(s) of Change (Split the PR if you check many)
- [ ] Added user feature
- [ ] Added technical feature
- [x] Added tests
- [ ] Refactor
- [ ] Bugfix
- [ ] Updated documentation
- [ ] Kubernetes changes
- [ ] CI/CD workflow changes

### Testing Checklist
- [ ] I have added/updated unit tests for functional changes.
- [ ] I have added/updated integration tests for changes that affect dependent systems.
- [x] I have added/updated end-to-end tests for feature changes.

### General Checklist
- [ ] I have pulled in the latest changes from master.
- [ ] I have run `make lint`.
- [ ] I have successfully run `make tests`.
- [ ] I have updated relevant CI/CD workflows.
- [ ] I have updated relevant Kubernetes manifests/scripts.
- [ ] I have updated/added relevant (internal and external) documentation (readme, doc comments, etc).
- [ ] I have updated the dependency map to reflect my changes.
- [ ] I have added the license to new source code files.
- [x] I have spent some time looking over the full diff before creating this PR.

### CODE REVIEW

File: lib/encryptonize.go [hunk 1]
```diff
@@ -251,14 +251,19 @@ func (e *Encryptonize) AuthorizeUser(user *SealedUser, access *SealedAccess) err
...
 	sealedGroup, err := (&group).seal(e.groupCryptor)
 	if err != nil {
 		return SealedUser{}, SealedGroup{}, "", err
 	}
 
-	user, pwd, err := newUser(append(groups, sealedGroup.ID)...)
+	groupIDs, err := e.verifyGroups(groups...)
+	if err != nil {
+		return SealedUser{}, SealedGroup{}, "", err
+	}
```
Comment (philip-cc) [inline lib/encryptonize.go:264]: We should probably do this as the first thing, so we can abort early if one of the groups isn't valid.

Comment (AmalieDue) [reply lib/encryptonize.go]: Done

File: lib/encryptonize_test.go [hunk 2]
```diff
@@ -61,3 +61,709 @@ func TestEncryptDecrypt(t *testing.T) {
...
+
+	_, access, err := enc.Encrypt(&user1, &plainObject)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = enc.AuthorizeUser(&user1, &access)
+	if err != nil {
+		t.Fatal(err)
+	}
```
Comment (philip-cc) [inline lib/encryptonize_test.go:397]: For stuff like this, you can also use the following syntax (might also be useful in other places in this file):
```suggestion
	if err := enc.AuthorizeUser(&user1, &access); err != nil {
		t.Fatal(err)
	}
```

File: lib/encryptonize_test.go [hunk 3]
```diff
@@ -61,3 +61,709 @@ func TestEncryptDecrypt(t *testing.T) {
...
+}
+
+// Scenario: 
+// 1) user1 is created.
+// 2) user1 creates two additional groups, group1 and group2.
+// 3) group2 is broken by changing some of its ciphertext.
+// 4) It is verified that user2 can be created and added to its own group and group1 simultaneously.
+// 5) It is verified that a third user cannot be created and added to its own group and group2 simultaneously (because of step 3) ).
+func TestNewUser(t *testing.T) {
```
Comment (philip-cc) [inline lib/encryptonize_test.go:423]: It seems like this tests both `NewUser` and `GetUserGroups`. Maybe it should be split into two tests?

File: lib/encryptonize_test.go [hunk 4]
```diff
@@ -61,3 +61,709 @@ func TestEncryptDecrypt(t *testing.T) {
...
+	}
+
+	_, _, _, err = enc.NewUser([]byte("data"), &group2)
+	if err == nil {
+		t.Fatal(err)
+	}
+}
+
+// It is verified that both empty and non-empty group data is accepted. Group data can be provided through either NewUser or NewGroup.
+func TestGroupData(t *testing.T) {
```
Comment (philip-cc) [inline lib/encryptonize_test.go:472]: This should probably be two tests: one for `NewUser` and one for `NewGroup`. Both tests should probably test that calling `GetGroupData` returns the expected results.

File: lib/encryptonize_test.go [hunk 5]
```diff
@@ -61,3 +61,709 @@ func TestEncryptDecrypt(t *testing.T) {
...
+
+	user2, _, _, err := enc.NewUser(nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	user3, _, _, err := enc.NewUser(nil)
+	if err != nil {
+		t.Fatal(err)
+	}
```
Comment (philip-cc) [inline lib/encryptonize_test.go:558]: You could consider doing this in a loop and storing the users in an array. Might be less verbose.

File: lib/encryptonize_test.go [hunk 6]
```diff
@@ -61,3 +61,709 @@ func TestEncryptDecrypt(t *testing.T) {
...
+
+	err = enc.AddUserToGroups(&user1, &user2, &group1, &group2)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	userGroups, err := enc.GetUserGroups(&user2)
```
Comment (philip-cc) [inline lib/encryptonize_test.go:699]: Unchecked error.

File: lib/encryptonize_test.go [hunk 7]
```diff
@@ -61,3 +61,709 @@ func TestEncryptDecrypt(t *testing.T) {
...
+	if _, ok := userGroups[group2.ID]; !ok {
+		t.Fatal("User not correctly added to group")
+	}
+
+	err = enc.RemoveUserFromGroups(&user1, &user2, &group1, &group2)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	userGroups, err = enc.GetUserGroups(&user2)
```
Comment (philip-cc) [inline lib/encryptonize_test.go:714]: Unchecked error.

File: lib/encryptonize_test.go [hunk 8]
```diff
@@ -61,3 +61,709 @@ func TestEncryptDecrypt(t *testing.T) {
...
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = enc.RemoveUserFromGroups(&user, &user, &group)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	userGroups, err := enc.GetUserGroups(&user)
```
Comment (philip-cc) [inline lib/encryptonize_test.go:759]: Unchecked error.

File: lib/encryptonize_test.go [hunk 9]
```diff
@@ -61,3 +61,709 @@ func TestEncryptDecrypt(t *testing.T) {
...
+	user.Ciphertext[2] = 0
+	user.Ciphertext[3] = 0
+	user.Ciphertext[4] = 0
+
+	plainObject := Object{
+		Plaintext:		[]byte("plaintext"),
+		AssociatedData:	[]byte("associated_data"),
+	}
+
+	_, _, err = enc.Encrypt(&user, &plainObject)
```
[the remark this answers is not in the document]
Comment (AmalieDue) [reply lib/encryptonize_test.go]: fixed

File: lib/encryptonize_test.go [hunk 10]
```diff
@@ -61,3 +61,709 @@ func TestEncryptDecrypt(t *testing.T) {
...
+	if err == nil {
+		t.Fatal(err)
```
[the remark this answers is not in the document]
Comment (AmalieDue) [reply lib/encryptonize_test.go]: fixed

File: lib/encryptonize_test.go [hunk 11]
```diff
@@ -61,3 +61,709 @@ func TestEncryptDecrypt(t *testing.T) {
...
+	}
+
+	decrypted, err := enc.Decrypt(&user1, &objectUpdated, &access)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if !reflect.DeepEqual(plainObjectUpdated, decrypted) {
+		t.Fatal("Decrypted object not equal to original")
```
[the remark this answers is not in the document]
Comment (AmalieDue) [reply lib/encryptonize_test.go]: fixed

File: lib/encryptonize_test.go [hunk 12]
```diff
@@ -61,3 +61,709 @@ func TestEncryptDecrypt(t *testing.T) {
...
+	}
+}
+
+// It is verified that plain objects with the following properties can be encrypted:
+// 1) Plaintext is empty
+// 2) Associated data is empty
+// 3) Both are empty
+// 4) Both are non-empty
+func TestPlainObject(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (AmalieDue) [reply lib/encryptonize_test.go]: fixed

File: lib/encryptonize_test.go (hunk 3 above)
Comment (AmalieDue) [reply lib/encryptonize_test.go:429]: fixed

File: lib/encryptonize_test.go (hunk 4 above)
Comment (AmalieDue) [reply lib/encryptonize_test.go]: fixed

File: lib/encryptonize_test.go [hunk 13]
```diff
@@ -61,3 +61,709 @@ func TestEncryptDecrypt(t *testing.T) {
...
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = enc.RemoveUserFromGroups(&user1, &user1, &group)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	userGroups, err := enc.GetUserGroups(&user1)
```
[the remark this answers is not in the document]
Comment (AmalieDue) [reply lib/encryptonize_test.go:702]: fixed

File: lib/encryptonize_test.go (hunk 7 above)
Comment (AmalieDue) [reply lib/encryptonize_test.go:769]: fixed

File: lib/encryptonize_test.go [hunk 14]
```diff
@@ -61,3 +61,765 @@ func TestEncryptDecrypt(t *testing.T) {
...
+		t.Fatal("Unauthorized user able to be authorized")
+	}
+}
+
+// Scenario: 
+// 1) user1 is created.
+// 2) user1 creates two additional groups, group1 and group2.
+// 3) group2 is broken by changing some of its ciphertext.
+// 4) It is verified that user2 can be created and added to its own group and group1 simultaneously.
+// 5) It is verified that a third user cannot be created and added to its own group and group2 simultaneously (because of step 3) ).
```
Comment (philip-cc) [inline lib/encryptonize_test.go:428]: This description doesn't seem to match the test.

File: lib/encryptonize_test.go [hunk 15]
```diff
@@ -61,3 +61,765 @@ func TestEncryptDecrypt(t *testing.T) {
...
+		t.Fatal(err)
+	}
+
+	copy(group2.Ciphertext[:5], make([]byte, 5))
+
+	if _, _, _, err = enc.NewUser([]byte("data"), &group1); err != nil {
+		t.Fatal(err)
+	}
+
+	if _, _, _, err = enc.NewUser([]byte("data"), &group2); err == nil {
```
Comment (philip-cc) [inline lib/encryptonize_test.go:453]: It would be good to check for leaked return values here.

File: lib/encryptonize_test.go [hunk 16]
```diff
@@ -61,3 +61,765 @@ func TestEncryptDecrypt(t *testing.T) {
...
+	group, err := enc.NewGroup(&user1, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if _, err = enc.GetGroupData(&user1, &group); err != nil {
+		t.Fatal(err)
+	}
+
+	if _, err = enc.GetGroupData(&user2, &group); err == nil {
```
Comment (philip-cc) [inline lib/encryptonize_test.go:681]: It would be good to check for leaked return values here.

**Review by philip-cc — COMMENTED** [review]
I think we're missing some tests around the sharing of objects. Some ideas:
* User1 creates an object, adds users2, verify that user2 can decrypt.
* User1 creates an object, adds users2. User2 removes user1, verify that user1 cannot decrypt.
* User1 creates object, adds user2, user2 adds user3, verify that user 3 can decrypt.
* Create a group with e.g. 5 users. Create and object and add the group. Verify that all group members can decrypt.

File: lib/encryptonize_test.go [hunk 17]
```diff
@@ -61,3 +61,765 @@ func TestEncryptDecrypt(t *testing.T) {
...
+// 1) Three users are created, user1, user2, and user3.
+// 2) user1 creates a group.
+// 3) It is verified that user1 is able to add and remove user2 from the group.
+// 4) It is verified that user3 who is not a member of the group is not able to add and remove user2 from the group.
+func TestAddRemoveUserFromGroups(t *testing.T) {
+	enc := newTestEncryptonize(t)
+
+	numUsers := 3
+	users := make([]SealedUser, 0, numUsers)
+	for i := 0 ; i < numUsers ; i++ {
```
[the remark this answers is not in the document]
Comment (AmalieDue) [reply lib/encryptonize_test.go]: Fixed

File: lib/encryptonize_test.go [hunk 18]
```diff
@@ -61,3 +61,765 @@ func TestEncryptDecrypt(t *testing.T) {
...
+	plainObject := Object{
+		Plaintext:		[]byte("plaintext"),
+		AssociatedData:	[]byte("associated_data"),
+	}
+
+	object, access, err := enc.Encrypt(&user, &plainObject)
+	if err == nil {
+		t.Fatal("Unauthorized user able to encrypt")
+	}
+	if object.Ciphertext != nil {
```
[the remark this answers is not in the document]
Comment (AmalieDue) [reply lib/encryptonize_test.go]: Fixed

File: lib/encryptonize_test.go [hunk 19]
```diff
@@ -61,3 +61,765 @@ func TestEncryptDecrypt(t *testing.T) {
...
+	object, access, err := enc.Encrypt(&user, &plainObject)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Make user unauthorized by changing its first 5 ciphertext bytes to 0 
+	copy(user.Ciphertext[:5], make([]byte, 5))
+
+	decryptedObject, err := enc.Decrypt(&user, &object, &access)
+	if err == nil || decryptedObject.Plaintext != nil {
```
[the remark this answers is not in the document]
Comment (AmalieDue) [reply lib/encryptonize_test.go]: Fixed

File: lib/encryptonize_test.go [hunk 20]
```diff
@@ -61,3 +61,765 @@ func TestEncryptDecrypt(t *testing.T) {
...
+
+	plainObjectUpdated := Object{
+		Plaintext:		[]byte("plaintext_updated"),
+		AssociatedData:	[]byte("associated_data_updated"),
+	}
+
+	// Make user unauthorized by changing its first 5 ciphertext bytes to 0 
+	copy(user.Ciphertext[:5], make([]byte, 5))
+
+	if _, err = enc.Update(&user, &plainObjectUpdated, &access); err == nil {
```
[the remark this answers is not in the document]
Comment (AmalieDue) [reply lib/encryptonize_test.go]: Fixed

File: lib/encryptonize_test.go [hunk 21]
```diff
@@ -61,3 +61,765 @@ func TestEncryptDecrypt(t *testing.T) {
...
+// 3) Both are empty
+// 4) Both are non-empty
+func TestPlainObject(t *testing.T) {
+	enc := newTestEncryptonize(t)
+
+	user, _, _, err := enc.NewUser(nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
```
[the remark this answers is not in the document]
Comment (AmalieDue) [reply lib/encryptonize_test.go:211]: Fixed

File: lib/encryptonize_test.go [hunk 22]
```diff
@@ -61,3 +61,765 @@ func TestEncryptDecrypt(t *testing.T) {
...
+	contents, err := enc.GetTokenContents(&token)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if !reflect.DeepEqual(plaintext, contents) {
+		t.Fatal("Token contents not equal to original")
+	}
```
[the remark this answers is not in the document]
Comment (AmalieDue) [reply lib/encryptonize_test.go]: Fixed

File: lib/encryptonize_test.go [hunk 23]
```diff
@@ -61,3 +61,765 @@ func TestEncryptDecrypt(t *testing.T) {
...
+		if _, _, err = enc.Encrypt(&user, &plainObject); err != nil {
+			t.Fatal(err)
+		}
+	})
+}
+
+// Scenario:
+// 1) A token is created.
+// 2) It is verified that the token contents is correctly derived.
+// 3) It is verified that if the token ciphertext is changed, then contents cannot be derived anymore.
```
[the remark this answers is not in the document]
Comment (AmalieDue) [reply lib/encryptonize_test.go]: Fixed

File: lib/encryptonize_test.go [hunk 24]
```diff
@@ -61,3 +61,765 @@ func TestEncryptDecrypt(t *testing.T) {
...
+	}
+
+	if !reflect.DeepEqual(plaintext, contents) {
+		t.Fatal("Token contents not equal to original")
+	}
+
+	// Change the first 5 token ciphertext bytes to 0 and verify that contents cannot be derived.
+	copy(token.Ciphertext[:5], make([]byte, 5))
+
+	contents, err = enc.GetTokenContents(&token)
```
[the remark this answers is not in the document]
Comment (AmalieDue) [reply lib/encryptonize_test.go]: Fixed

File: lib/encryptonize_test.go [hunk 25]
```diff
@@ -61,3 +61,765 @@ func TestEncryptDecrypt(t *testing.T) {
...
+
+	accessGroups, err := enc.GetAccessGroups(&user1, &access)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if _, ok := accessGroups[user1.ID]; !ok {
+		t.Fatal("User's own group not added to access object encrypted by user")
+	}
+
+	if _, err = enc.GetAccessGroups(&user2, &access); err == nil {
```
[the remark this answers is not in the document]
Comment (AmalieDue) [reply lib/encryptonize_test.go]: Fixed

File: lib/encryptonize_test.go [hunk 26]
```diff
@@ -61,3 +61,765 @@ func TestEncryptDecrypt(t *testing.T) {
...
+	if err = enc.AddGroupsToAccess(&user1, &access, &group); err != nil {
+		t.Fatal(err)
+	}
+
+	accessGroups, err := enc.GetAccessGroups(&user1, &access)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if _, ok := accessGroups[group.ID]; !ok {
```
[the remark this answers is not in the document]
Comment (AmalieDue) [reply lib/encryptonize_test.go]: Fixed

File: lib/encryptonize_test.go (hunk 14 above)
Comment (AmalieDue) [reply lib/encryptonize_test.go]: Fixed

File: lib/encryptonize_test.go (hunk 15 above)
Comment (AmalieDue) [reply lib/encryptonize_test.go]: Fixed

File: lib/encryptonize_test.go (hunk 16 above)
Comment (AmalieDue) [reply lib/encryptonize_test.go]: Fixed

File: lib/encryptonize_test.go [hunk 27]
```diff
@@ -56,8 +57,1023 @@ func TestEncryptDecrypt(t *testing.T) {
...
+	group, err := enc.NewGroup(&users[0], []byte("data"))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	for i := 1; i < numUsers; i++ {
+		if err = enc.AddUserToGroups(&users[0], &users[i], &group); err != nil {
+			t.Fatal(err)
+		}
+	}
```
Comment (philip-cc) [inline lib/encryptonize_test.go:1058]:
```suggestion
	for i, _ := range users {
		if err = enc.AddUserToGroups(&users[0], &users[i], &group); err != nil {
			t.Fatal(err)
		}
	}
```

File: lib/encryptonize_test.go [hunk 28]
```diff
@@ -56,8 +57,1023 @@ func TestEncryptDecrypt(t *testing.T) {
...
+
+	if err = enc.AddGroupsToAccess(&users[0], &access, &group); err != nil {
+		t.Fatal(err)
+	}
+
+	for i := 0; i < numUsers; i++ {
+		if _, err = enc.Decrypt(&users[i], &object, &access); err != nil {
+			t.Fatal(err)
+		}
+	}
```
Comment (philip-cc) [inline lib/encryptonize_test.go:1078]:
```suggestion
	for _, user := range users {
		if _, err = enc.Decrypt(&user, &object, &access); err != nil {
			t.Fatal(err)
		}
	}
```

**Review by philip-cc — COMMENTED** [review]
If you run `go test -coverprofile=coverage.out -count=1 ./... && go tool cover -html=coverage.out` you can see that:
* We never hit a failure in `verifyGroups` in `AddGroupsToAccess` and `RemoveGroupsFromAccess`.
* We never pass an invalid user in `GetUserGroups`, `AuthenticateUser`, `AddUserToGroups`, `RemoveUserFromGroups`, `NewGroup`
* we never pass invalid input to `GetGroupData`

Otherwise it's starting to look good.

**Comment by AmalieDue** [comment]
Fixed (added more tests with invalid user/group).

**Review by philip-cc — APPROVED** [review]
Ship it!

**Pull request merged**